src/metabase/db/liquibase.clj

```diff
@@ -216,66 +219,92 @@
-     (wait-for-migration-lock-to-be-cleared liquibase)
-     ;; while we were waiting for the lock, it was possible that another instance finished the migration(s), so make
-     ;; sure something still needs to be done...
-     (let [to-run-migrations      (unrun-migrations data-source)
-           unrun-migrations-count (count to-run-migrations)]
-       (if (pos? unrun-migrations-count)
-         (let [^Contexts contexts nil
-               start-time         (System/currentTimeMillis)]
-           (log/info (trs "Running {0} migrations ..." unrun-migrations-count))
-           (doseq [^ChangeSet change to-run-migrations]
-             (log/tracef "To run migration %s" (.getId change)))
-           (.update liquibase contexts)
-           (log/info (trs "Migration complete in {0}" (u/format-milliseconds (- (System/currentTimeMillis) start-time)))))
-         (log/info
-          (trs "Migration lock cleared, but nothing to do here! Migrations were finished by another instance.")))))
-    (log/info (trs "No unrun migrations found."))))
+(defn- assert-locked [liquibase]
+  (when-not (.hasChangeLogLock (lock-service liquibase))
+    ;; TODO not sure this is the right exception class to use
+    (throw (LockException. "The Liquibase lock must be taken before executing this operation"))))
+
+(def ^:private ^:dynamic *lock-depth* 0)
 
 (defn run-in-scope-locked
   "Run function `f` in a scope on the Liquibase instance `liquibase`.
-  Liquibase scopes are used to hold configuration and parameters (akin to binding dynamic variables in
-  Clojure). This function initializes the database and the resource accessor which are often required."
+   Liquibase scopes are used to hold configuration and parameters (akin to binding dynamic variables in
+   Clojure). This function initializes the database and the resource accessor which are often required.
+   The underlying locks are re-entrant, so it is safe to nest these blocks."
   [^Liquibase liquibase f]
-  (let [database (.getDatabase liquibase)
-        ^LockService lock-service (.getLockService (LockServiceFactory/getInstance) database)
-        scope-objects {(.name Scope$Attr/database) database
+
+  (when (.hasChangeLogLock (lock-service liquibase))
+    ;; In production it's better to just take the lock again, it is re-entrant.
+    ;; Fail in dev and CI in order to clean up the code flow.
+    (when-not config/is-prod?
+      (throw (LockException. "Attempted to take a Liquibase lock, but we already are holding it."))))
+
+  (let [database      (.getDatabase liquibase)
+        scope-objects {(.name Scope$Attr/database)         database
                        (.name Scope$Attr/resourceAccessor) (.getResourceAccessor liquibase)}]
     (Scope/child ^Map scope-objects
                  (reify Scope$ScopedRunner
                    (run [_]
-                     (.waitForLock lock-service)
+                     (wait-for-migration-lock liquibase)
+                     (assert-locked liquibase)
```

Comment:
why do you need to assert here?

Reply:
Oh this is left over from debugging - the `.clearCheckSums` call was clearing the lock so I put these statements all over the place to find where that was happening. Will remove.